targets/pna_nic/pna_nic.cpp

```diff
@@ -204,6 +204,21 @@ PnaNic::main_thread() {
 
     Deparser *deparser = this->get_deparser("main_deparser");
     deparser->deparse(packet.get());
+
+    // recirculation
+    if ((port_t) packet->get_egress_port() == PNA_PORT_RECIRCULATE) {
+      BMLOG_DEBUG_PKT(*packet, "Recirculating packet");
+
+      phv->reset();
+      phv->reset_header_stacks();
+      phv->reset_metadata();
+
+      phv->get_field("pna_main_parser_input_metadata.recirculated")
+        .set(1);
+      input_buffer.push_front(std::move(packet));
```

Comment:
I believe it should be unchanged from the original.  Unfortunately some things like this are not yet spelled out in as precise a form in the PNA architecture as they are in the PSA architecture, but I believe "input_port is unchanged after recirculate operation" is a reasonable starting place (and maybe finishing place, too).

Reply:
Thanks. This is not what the current PR does, but it can be done easily. See PSA implementation: https://github.com/p4lang/behavioral-model/blob/28b736caabca7a4c9ff23e3b947354bd55191372/targets/psa_switch/psa_switch.cpp#L618-L619

Reply:
Ok. Will update the line.

Just a doubt: Why are we setting `ingress_port` field to `PSA_PORT_RECIRCULATE` ? The P4 programs will find whether the packet is recirculated on not using `recirculated_flag` field.

Reply:
According to @jafingerhut, the port should be set to the original ingress port, not `PSA_PORT_RECIRCULATE`
But we should still set the port to something. With your current implementation, the port will always be set to 0. At least that's what will happen I think based on the `reset_metadata` call (I didn't test).

Reply:
One approach is to wait until there is an official PNA definition for how recirculation is done, and behaves, and implement other features that are already well-defined.